### PR TITLE
Feat(UsingWitnet): self-estimate reward values when posting or upgrading requests

### LINKED ISSUE #193: Assess variants to `UsingWitnet._witnetPostRequest(..)`
As current implementation will take the whole `msg.value` when internally calling to the WRB.

### PR DESCRIPTION
**Pull request by guidiaz**
Contributes to #193.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -190,6 +190,20 @@ From the point of view of a `UsingWitnet` contract, any given `WitnetRequestBoar
 
 The `UsingWitnet` contract injects the following _internal methods_ into the contracts inheriting from it:
 
+- **`_witnetEstimateReward()`**:
+  - _Description_:
+    - Estimates the reward amount, considering current transaction gas price.
```
Comment (aesedepece) [inline README.md:195]:
```suggestion
    - Estimates the reward amount, considering current transaction's gas price.
```

File: README.md [hunk 2]
```diff
@@ -190,6 +190,20 @@ From the point of view of a `UsingWitnet` contract, any given `WitnetRequestBoar
...
+  - _Returns_:
+    - The reward to be included when either posting a new request, or upgrading the reward of a previously posted one.
+
+- **`_witnetEstimateReward(uint256 _gasPrice)`**:
+  - _Description_:
+    - Estimates the reward amount, considering the give gas price is used when either posting or upgrading a request.
```
Comment (aesedepece) [inline README.md:201]:
```suggestion
    - Estimates the reward amount, considering the given gas price is used when either posting or upgrading a request.
```

File: contracts/UsingWitnet.sol [hunk 3]
```diff
@@ -37,45 +37,73 @@ abstract contract UsingWitnet {
...
+        returns (uint256)
+    {
+        return witnet.estimateReward(tx.gasprice);
+    }
+
     /// Send a new request to the Witnet network with transaction value as a reward.
     /// @param _request An instance of `IWitnetRequest` contract.
-    /// @return Sequential identifier for the request included in the WitnetRequestBoard.
+    /// @return _id Sequential identifier for the request included in the WitnetRequestBoard.
+    /// @return _actualReward Actual reward amount escrowed by the WRB until a result gets reported.
```
Comment (mariocao) [inline contracts/UsingWitnet.sol:70]:
```suggestion
    /// @return _reward Current reward amount escrowed by the WRB until a result gets reported.
```

File: contracts/UsingWitnet.sol [hunk 4]
```diff
@@ -37,45 +37,73 @@ abstract contract UsingWitnet {
...
     function _witnetPostRequest(IWitnetRequest _request)
         internal
-        returns (uint256)
+        virtual
+        returns (uint256 _id, uint256 _actualReward)
     {
-        return witnet.postRequest{value: msg.value}(_request);
+        _actualReward = _witnetEstimateReward();
```
Comment (mariocao) [inline contracts/UsingWitnet.sol:76]:
```suggestion
        _reward = _witnetEstimateReward();
 ```

File: contracts/UsingWitnet.sol [hunk 5]
```diff
@@ -37,45 +37,73 @@ abstract contract UsingWitnet {
...
+        _id = witnet.postRequest{value: _actualReward}(_request);
```
Comment (mariocao) [inline contracts/UsingWitnet.sol:77]:
```suggestion
        _id = witnet.postRequest{value: _reward}(_request);
```

File: test/helpers/UsingWitnetTestHelper.sol [hunk 6]
```diff
@@ -19,17 +19,25 @@ contract UsingWitnetTestHelper is UsingWitnet {
...
     external payable
-    returns(uint256 id)
+    returns(uint256 _id)
   {
-    return _witnetPostRequest(_script);
+    uint256 _actualReward;
+    (_id, _actualReward) = _witnetPostRequest(_request);
+    if (_actualReward < msg.value) {
+      payable(msg.sender).transfer(msg.value - _actualReward);
+    }
```
Comment (mariocao) [inline test/helpers/UsingWitnetTestHelper.sol:30]: Simply use `_reward`

File: contracts/UsingWitnet.sol (hunk 3 above)
Comment (guidiaz) [reply contracts/UsingWitnet.sol:70]: With "actual" I refer to the "real" reward sent to the WRB, which would be less than the `msg.value` sent by the caller in most cases. However, I agree that referring just to "reward" could be easier to follow.

File: contracts/UsingWitnet.sol [hunk 7]
```diff
@@ -37,45 +37,73 @@ abstract contract UsingWitnet {
...
+        uint256 _currentReward = witnet.readRequestReward(_id);        
+        uint256 _newReward = _witnetEstimateReward();
+        if (_newReward > _currentReward) {
+            uint256 _fundsToAdd = (_newReward - _currentReward);
+            witnet.upgradeReward{value: _fundsToAdd}(_id);
+            return _fundsToAdd;
+        } else {
+            witnet.upgradeReward{value: 0}(_id); // Let Request.gasPrice be updated.
+            return 0;
+        }
```
Comment (aesedepece) [inline contracts/UsingWitnet.sol:98]: To avoid code repetition and to honor the "single exit point" principle, let me suggest:
```suggestion
        uint256 _fundsToAdd = 0;
        if (_newReward > _currentReward) {
            _fundsToAdd = (_newReward - _currentReward);
        }
        witnet.upgradeReward{value: _fundsToAdd}(_id); // Let Request.gasPrice be updated
        return _fundsToAdd;
```

File: package.json [hunk 8]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "witnet-ethereum-bridge",
-  "version": "0.4.3",
+  "version": "0.5.0",
```
Comment (aesedepece) [inline package.json:3]: As a rule of thumb, version bumps should go into their own commit at the end of the PR, clearly stating `chore: bump package version to 0.5.0`

**Review by aesedepece — CHANGES_REQUESTED** [review]

**Review by aesedepece — APPROVED** [review]

**Comment by aesedepece** [comment]
Let's get this merged!
